src/i2c/src/i2c.rs

```diff
@@ -0,0 +1,548 @@
+            },
+        })
+    }
+
+    /// Sets client's address for an I2C adapter.
+    fn set_client_addr(&self, addr: usize, dryrun: bool) -> Result<()> {
+        if dryrun {
+            return Ok(());
+        }
+
+        let ret = unsafe { ioctl(self.fd.as_raw_fd(), I2C_SLAVE, addr as c_ulong) };
+
+        if ret == -1 {
+            println!("Failed to set client addr to {:x}", addr);
+            errno_result()
+        } else {
+            Ok(())
+        }
+    }
+
+    /// Gets adapter's functionality
+    fn get_func(&mut self, dryrun: bool) -> Result<()> {
+        let func: u64 = I2C_FUNC_SMBUS_ALL;
+
+        if !dryrun {
+            let ret = unsafe { ioctl(self.fd.as_raw_fd(), I2C_FUNCS, &func) };
+
+            if ret == -1 {
+                println!("Failed to get I2C function");
+                return errno_result();
+            }
+        }
+
+        if (func & I2C_FUNC_I2C) != 0 {
+            self.smbus = false;
+        } else if (func & I2C_FUNC_SMBUS_ALL) != 0 {
+            self.smbus = true;
+        } else {
+            println!("Invalid functionality {:x}", func);
+            return Err(Error::new(EINVAL));
+        }
+
+        Ok(())
+    }
+
+    /// Transfer data
+    fn do_i2c_transfer(&self, data: &I2cRdwrIoctlData, addr: u16) -> Result<()> {
+        let ret = unsafe { ioctl(self.fd.as_raw_fd(), I2C_RDWR, data) };
+
+        if ret == -1 {
+            println!("Failed to transfer i2c data to client addr to {:x}", addr);
+            errno_result()
+        } else {
+            Ok(())
+        }
+    }
+
+    fn do_smbus_transfer(&self, data: &I2cSmbusIoctlData, addr: u16, dryrun: bool) -> Result<()> {
+        if !dryrun {
+            let ret = unsafe { ioctl(self.fd.as_raw_fd(), I2C_SMBUS, data) };
+
+            if ret == -1 {
+                println!("Failed to transfer smbus data to client addr to {:x}", addr);
+                return errno_result();
+            }
+        }
+
+        Ok(())
+    }
+
+    /// Perform I2C_RDWR transfer
+    fn i2c_transfer(&self, reqs: &mut [I2cReq]) -> Result<()> {
+        let mut msgs: Vec<I2cMsg> = Vec::with_capacity(reqs.len());
+        let len = reqs.len();
+        let addr = reqs[0].addr;
+
+        for req in reqs {
+            msgs.push(I2cMsg {
+                addr: req.addr,
+                flags: req.flags,
+                len: req.len,
+                buf: req.buf.as_mut_ptr(),
+            });
+        }
+
+        let data = I2cRdwrIoctlData {
+            msgs: msgs.as_mut_ptr(),
+            nmsgs: len as u32,
+        };
+
+        self.do_i2c_transfer(&data, addr)
+    }
+
+    /// Based on Linux's drivers/i2c/i2c-core-smbus.c:i2c_smbus_xfer_emulated().
+    ///
+    /// These smbus related functions try to reverse what Linux does, only support basic modes
+    /// (up to word transfer).
+    fn smbus_prepare(&self, reqs: &mut [I2cReq]) -> Result<I2cSmbusIoctlData> {
+        let mut data = I2cSmbusData {
```

Comment:
How about adding a helper I2cSmbusData::new() for initializaiton?

Reply:
> How about adding a helper I2cSmbusData::new() for initializaiton?

Sure, that's better.